velox/functions/lib/Re2Functions.cpp

```diff
@@ -970,89 +970,287 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> re2ExtractSignatures() {
+      if (currentChar != '%' && currentChar != '_' &&
+          currentChar != escapeChar_) {
+        VELOX_USER_FAIL(
+            "Escape character must be followed by '%', '_' or the escape character itself: {}, escape {}",
+            pattern_,
+            escapeChar_.value())
+      }
+    }
+
+    return true;
+  }
+
+  // Current index of the cursor.
+  char currentIndex() const {
+    return currentIndex_;
+  }
+
+  bool isAnyCharsWildcard() const {
+    return charKind_ == CharKind::kAnyCharsWildcard;
+  }
+
+  bool isSingleCharWildcard() const {
+    return charKind_ == CharKind::kSingleCharWildcard;
+  }
+
+  bool isWildcard() {
+    return isAnyCharsWildcard() || isSingleCharWildcard();
+  }
+
+  bool isPreviousWildcard() {
+    return userPreviousCharKind_ == CharKind::kAnyCharsWildcard ||
+        userPreviousCharKind_ == CharKind::kSingleCharWildcard;
+  }
+
+ private:
+  // Represents the state of current cursor/char.
+  enum class CharKind {
+    // Escape char.
+    // NOTE: If escape char is set as '\',  for pattern '\\', the first '\' is
+    // an escaping char, the second is not, it is just a literal '\'
+    kEscape,
+    // Wildcard char: %.
+    // NOTE: If escape char is set as '\', for pattern '\%%', the first '%' is
+    // not a wildcard, just a literal '%', the second '%' is a wildcard.
+    kAnyCharsWildcard,
+    // Wildcard char: _.
+    // NOTE: If escape char is set as '\', for pattern '\__', the first '_' is
+    // not a wildcard, just a literal '_', the second '_' is a wildcard.
+    kSingleCharWildcard,
+    // Chars that are not escape char & not wildcard char.
+    kNormal
+  };
+
+  // Advance the cursor to next char.
+  void nextInternal() {
+    const bool previousEscape = (charKind_ == CharKind::kEscape);
+
+    currentIndex_++;
+    auto currentChar = current();
+    if (!previousEscape && currentChar == escapeChar_) {
+      charKind_ = CharKind::kEscape;
+    } else if (
+        !previousEscape && currentChar != escapeChar_ &&
+        (currentChar == '_' || currentChar == '%')) {
+      if (currentChar == '_') {
+        charKind_ = CharKind::kSingleCharWildcard;
+      } else {
+        charKind_ = CharKind::kAnyCharsWildcard;
+      }
+    } else {
+      charKind_ = CharKind::kNormal;
+    }
+  }
+
+  // Char at current cursor.
+  char current() const {
+    return pattern_.data()[currentIndex_];
+  }
+
+  const StringView pattern_;
+  const std::optional<char> escapeChar_;
+  const size_t lastIndex_;
+
+  int32_t currentIndex_ = -1;
+  // Kind of current char.
+  CharKind charKind_ = CharKind::kNormal;
+  // 'previousCharKind_' from user's perspective.
+  // The difference between previousCharKind_ and realPreviousCharKind_ can be
+  // described by an example: for pattern string 'a\_b', if the cursor is at '_'
+  // previousCharKind_ is kEscape, while userPreviousCharKind_ is kNormal.
+  CharKind userPreviousCharKind_ = CharKind::kNormal;
```

Comment:
It seems we can simplify some more. We can replace this with a bool isPreviousWildcard_; remove nextInternal; and remove CharKind::kEscape.

See last commit in https://github.com/facebookincubator/velox/pull/7933